cpp/test/sg/fil_test.cu

```diff
@@ -117,6 +124,39 @@ __global__ void nan_kernel(float* data, const bool* mask, int len, float nan)
 
 float sigmoid(float x) { return 1.0f / (1.0f + expf(-x)); }
 
+void hard_clipped_bernoulli(
+  raft::random::Rng rng, float* d, size_t n_vals, float prob_of_zero, cudaStream_t stream)
+{
+  rng.uniform(d, n_vals, 0.0f, 1.0f, stream);
+  thrust::transform(
+    thrust::cuda::par.on(stream), d, d + n_vals, d, [=] __device__(float uniform_0_1) {
+      float truly_0_1 = fmin(fmax(uniform_0_1, 0.0), 1.0);
+      return truly_0_1 > prob_of_zero;
+    });
+}
+
+struct replace_some_floating_with_categorical {
+  float* max_matching_cat_d;
+  int num_cols;
+  __device__ float operator()(float data, int data_idx)
+  {
+    int max_matching_cat = max_matching_cat_d[data_idx % num_cols];
+    if (max_matching_cat == -1.0) return data;
```

Comment:
```suggestion
    if (max_matching_cat == -1.0f) return data;
```

Reply:
will it matter in a memory-bound kernel? I was trying to make the code easier to read